engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java

```diff
@@ -890,6 +894,42 @@ public IdExecutionService getIdValueExtractor() {
     }
   }
 
+
+  /**
+   * Helper method to use when an unexpected state happens that should raise a panic,
+   * but not crash the interpreter. Creates a {@link PanicException} with
+   * <em>assertion error</em> payload.
+   *
+   *
+   * @param node where the problem happened (may be {@code null})
+   * @param message {@code null} (then {@code e.getMessage()} is used) or a special
+   *   message to use in the panic
+   * @param e external exception to extract message and stack from or {@code null}
+   * @return this method never returns it throws the {@link PanicException}
+   * @throws PanicException with <em>assertion error</em> payload
+   */
+  @CompilerDirectives.TruffleBoundary
+  public PanicException raiseAssertionPanic(Node node, String message, Throwable e) throws PanicException {
```

Comment:
I'd like us to use this method to throw in case an unexpected behavior rather than using `new IllegalStateException` (which cannot work on fast path) or `shouldNotReachHere` (unless justified). This method just propagates a standard guest `PanicException` and does not crash the interpreter (as the other methods do).

OK, @hubertp @Akirathan?

Reply:
For many places, this seems, indeed, as a more reasonable choice. However, I still think we have few places that should crash the interpreter.

Reply:
Right, I'd suggest:
- when the error shalln't happen, but may (for example by buggy `InteropLibrary` contract in another language) - use `raiseAssertionPanic`
- when the error cannot provably happen (if it does it is a sign of an error in Enso own code) use `CompilerDirectives.shouldNotReachHere`
- never throw exception that `fillInStackTrace` - e.g. regular Java exceptions